File: services/route/varuna_route/forecast.py
"""The depth a route reads off a run, per segment and per five-minute step (task P8.1).

**Read from ``segments_wet.json``, not from the parquet.** ``segment_forecast.parquet`` is the
product of record - 766,656 rows, every segment at every step - and re-reading 19 MB of it per
request would spend the whole 300 ms route budget on IO. The cycle also writes the compact form
this loads: only the segments the run ever wetted (1,498 of 21,296 on the 2 July storm), which is
the same information for routing purposes because a segment that never reaches 5 cm never changes
anybody's route.

**What the probabilities are worth (corrected 2026-09-19, task D-01).** This docstring used to
say "a baked run is deterministic - ``ensemble_n`` is 1", and the router computed
``P(h > threshold)`` as a hard 1 or 0 from the median depth. That stopped being true at the
2026-09-13 re-bake, which put 20 members through the emulator and wrote each wet segment's
``p_gt`` series at 15/30/45/60 cm into ``segments_wet.json`` (``write_wet_segments``). Every
probability the router printed was therefore 1.0 or 0.0 over a run that knew better, and
``risk_tolerance`` had nothing to bite on - it was a placebo.

So :meth:`SegmentDepths.exceedance` reads ``p_gt`` when the run carries it and falls back to the
median-depth step function only when it does not (a one-member run, or a bake older than that
change). Which of the two answered is on the response as a note, not buried: a probability that
is really a threshold comparison must not be read as an ensemble's opinion (CLAUDE.md 6).
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from functools import lru_cache
from pathlib import Path

import structlog
from varuna_schemas.paths import run_dir, runs_dir

log = structlog.get_logger("varuna.route.forecast")

__all__ = ["SegmentDepths", "latest_run_dir", "load_depths"]

STEP_MIN = 5
"""Forecast step, minutes. Matches the cycle's own step (CLAUDE.md 10.3)."""

DRY_CM = 5.0
"""Below this a segment is dry enough that no profile slows for it (CLAUDE.md 6.2 `--depth-dry`)."""


@dataclass(frozen=True, slots=True)
class SegmentDepths:
    """One run's depth series for every segment it wetted."""

    run_id: str
    times: tuple[datetime, ...]
    """The instant each forecast step is valid at, as the cycle wrote them."""

    depth_cm: dict[str, list[float]]
    n_steps: int
    n_total: int
    """Segments in the city, wet or not - so the response can say 1,498 of 21,296."""

    ensemble_n: int

    rain_aoi_mm_h: tuple[float, ...] = ()
    """The run's AOI-mean rain per forecast step, straight off ``run.json``.

    Carried here so a route can say what this cycle's rain peaks at beside what the drain under
    the street was designed for (PRD 3.4). It is an **AOI mean**, not the rain over one junction,
    and every response that quotes it says so."""

    p_gt: dict[float, dict[str, list[float]]] = field(default_factory=dict)
    """``{threshold_cm: {segment_id: [P(h > threshold) per step]}}``, empty when the run has none.

    Exactly the ``p_gt`` block ``segments_wet.json`` carries: the four thresholds of the depth
    ramp, over the segments the run wetted. The cycle writes it only when some value lies
    strictly between 0 and 1, so its presence is itself the claim that this run measured a
    spread."""

    @property
    def has_exceedance(self) -> bool:
        """True when the run carries per-member exceedance rather than a threshold comparison."""
        return bool(self.p_gt)

    @property
    def valid_ts(self) -> datetime:
        """When the forecast starts."""
        return self.times[0]

    def step_at(self, when: datetime) -> int:
        """Which forecast step covers an instant, clamped to the run's window.

        Clamped rather than refused: a route that runs past the end of the forecast is a real
        request ("leave at 09:30, the forecast ends at 09:40"), and holding the last step is the
        honest answer - the water at +180 min is the last thing this run knows.
        """
        delta = (when - self.times[0]).total_seconds() / 60.0
        return max(0, min(self.n_steps - 1, int(delta // STEP_MIN)))

    def depth_at(self, segment_id: str, step: int) -> float:
        """Depth in cm on a segment at a step; 0 for anything the run never wetted."""
        series = self.depth_cm.get(segment_id)
        if not series:
            return 0.0
        return series[step] if step < len(series) else series[-1]

    def exceedance(self, segment_id: str, threshold_cm: float, step: int) -> float:
        """``P(depth > threshold_cm)`` on a segment at a step.

        From the run's own ``p_gt`` where it has one - which is where ``risk_tolerance`` gets
        something to weigh - and otherwise from the median depth, which makes it 1 or 0 by
        construction. A segment the run never wetted is absent from both and cannot exceed
        anything, so it scores 0.
        """
        by_segment = self.p_gt.get(float(threshold_cm))
        if by_segment is not None:
            series = by_segment.get(segment_id)
            if series is not None:
                return series[step] if step < len(series) else series[-1]
            if segment_id not in self.depth_cm:
                return 0.0
        return 1.0 if self.depth_at(segment_id, step) > threshold_cm else 0.0

    def peak(self, segment_id: str) -> float:
        series = self.depth_cm.get(segment_id)
        return max(series) if series else 0.0

    def time_of(self, step: int) -> datetime:
        if step < len(self.times):
            return self.times[step]
        return self.times[-1] + timedelta(minutes=STEP_MIN * (step - len(self.times) + 1))


def latest_run_dir(city: str = "mumbai") -> Path:
    """The newest run **for a city** that carries a segment forecast.

    The city filter is load-bearing. Runs from every city share `data/runs/` and their ids sort
    chronologically, so once Chennai was onboarded its `CHN-` runs sorted above Mumbai's and the
    router began planning Mumbai trips against Chennai depths - a wrong answer that looked
    entirely normal.

    Raises:
        FileNotFoundError: nothing is baked. The message carries the command that fixes it.
    """
    from varuna_schemas.models.run import RunIdError, city_code

    try:
        prefix = f"{city_code(city)}-"
    except RunIdError:
        prefix = ""
    root = runs_dir()
    candidates = (
        [
            p
            for p in sorted(root.iterdir(), reverse=True)
            if p.is_dir()
            and (not prefix or p.name.startswith(prefix))
            and (p / "segments_wet.json").is_file()
        ]
        if root.is_dir()
        else []
    )
    if not candidates:
        msg = (
            "No baked run to route against. Press Play on the replay, or run "
            "`make bake BUNDLE=MUM-2019-07-02`."
        )
        raise FileNotFoundError(msg)
    return candidates[0]


@lru_cache(maxsize=4)
def _load(path_str: str, mtime_ns: int) -> SegmentDepths:
    del mtime_ns  # part of the cache key: a re-baked run invalidates itself.
    path = Path(path_str)
    wet = json.loads((path / "segments_wet.json").read_text(encoding="utf-8"))
    depth = {str(k): [float(v) for v in series] for k, series in wet.get("depth_cm", {}).items()}
    n_steps = max((len(v) for v in depth.values()), default=1)
    # `valid_ts` is the list of step times, one per forecast step, exactly as the cycle wrote it.
    stamps = wet.get("valid_ts") or []
    times = tuple(datetime.fromisoformat(str(t)) for t in stamps)
    if not times:
        msg = f"{path.name}/segments_wet.json carries no step times; it cannot be routed against."
        raise ValueError(msg)

    ensemble_n = 1
    rain: tuple[float, ...] = ()
    meta_path = path / "run.json"
    if meta_path.is_file():
        meta = json.loads(meta_path.read_text(encoding="utf-8"))
        ensemble_n = int(meta.get("ensemble_n", 1) or 1)
        rain = tuple(float(v) for v in (meta.get("rain_aoi_mm_h") or []))

    # `{"15": {segment_id: [...]}}` on the wire; keyed by float here so a profile's threshold
    # (15.0, 30.0, 45.0, 60.0 cm - the depth ramp) looks itself up without string formatting.
    p_gt = {
        float(threshold): {
            str(sid): [float(v) for v in series] for sid, series in by_segment.items()
        }
        for threshold, by_segment in (wet.get("p_gt") or {}).items()
    }

    depths = SegmentDepths(
        run_id=str(wet.get("run_id", path.name)),
        times=times,
        depth_cm=depth,
        n_steps=n_steps,
        n_total=int(wet.get("n_segments_total", len(depth))),
        ensemble_n=ensemble_n,
        rain_aoi_mm_h=rain,
        p_gt=p_gt,
    )
    log.info(
        "route.depths_loaded",
        run_id=depths.run_id,
        wet=len(depth),
        total=depths.n_total,
        steps=n_steps,
        ensemble_n=ensemble_n,
        p_gt=sorted(p_gt),
    )
    return depths


def load_depths(run_id: str | None = None, city: str = "mumbai") -> SegmentDepths:
    """Load a run's segment depths, cached on the file's mtime."""
    path = run_dir(run_id) if run_id else latest_run_dir(city)
    wet = path / "segments_wet.json"
    if not wet.is_file():
        msg = f"Run {path.name} has no segment forecast; it cannot be routed against."
        raise FileNotFoundError(msg)
    return _load(str(path), wet.stat().st_mtime_ns)

File: services/route-rs/src/forecast.rs
//! A run's depth per segment and step (`varuna_route.forecast`), read from the run's own files.
//!
//! Like the Python router this reads `segments_wet.json` (the wet segments' depth series and the
//! 20-member `p_gt` exceedance series) and `run.json` (`ensemble_n`, the AOI hyetograph), never
//! the 19 MB parquet. The series are re-keyed from segment id to the graph's segment index once,
//! at load, so the search indexes a vector where Python looks up a dict.

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use std::time::SystemTime;

use serde::Deserialize;
use serde_json::Value;

use crate::graph::RoadGraph;
use crate::pynum::{floordiv, str_value, trunc_i64};
use crate::pytime::{fromisoformat, DateTime, Parsed};

/// Forecast step, minutes (`forecast.STEP_MIN`).
pub const STEP_MIN: f64 = 5.0;
/// The same step in seconds (`forecast._STEP_S`).
pub const STEP_S: f64 = STEP_MIN * 60.0;
/// Below this a segment is dry (`forecast.DRY_CM`).
pub const DRY_CM: f64 = 5.0;

/// Per-segment series, indexed by the graph's segment index. `None` is "absent from the file",
/// which the Python lookups distinguish from an empty list.
pub type Series = Vec<Option<Vec<f64>>>;

pub struct SegmentDepths {
    pub run_id: String,
    pub times: Vec<DateTime>,
    pub depth: Series,
    pub n_steps: usize,
    pub ensemble_n: i64,
    pub rain_aoi_mm_h: Vec<f64>,
    /// `(threshold_cm, series)` for each threshold the run carries a `p_gt` block for.
    pub p_gt: Vec<(f64, Series)>,
    /// Whether the file had a `p_gt` block with at least one threshold (`has_exceedance`).
    pub has_exceedance: bool,
}

#[derive(Deserialize)]
struct WetFile {
    #[serde(default)]
    run_id: Option<Value>,
    #[serde(default)]
    valid_ts: Option<Vec<Value>>,
    #[serde(default)]
    depth_cm: Option<HashMap<String, Vec<f64>>>,
    #[serde(default)]
    p_gt: Option<HashMap<String, HashMap<String, Vec<f64>>>>,
}

#[derive(Deserialize)]
struct RunFile {
    #[serde(default)]
    ensemble_n: Option<Value>,
    #[serde(default)]
    rain_aoi_mm_h: Option<Vec<f64>>,
}

impl SegmentDepths {
    pub fn valid_ts(&self) -> DateTime {
        self.times[0]
    }

    /// `step_at(when)`: the step covering an instant, clamped to the window.
    pub fn step_at(&self, when: DateTime) -> usize {
        let delta = when.seconds_since(self.times[0]) / 60.0;
        let step = trunc_i64(floordiv(delta, STEP_MIN));
        let last = self.n_steps as i64 - 1;
        step.min(last).max(0) as usize
    }

    /// `depart_offset_s(when)`.
    pub fn depart_offset_s(&self, when: DateTime) -> f64 {
        when.seconds_since(self.times[0])
    }

    /// `step_after(depart_offset_s, seconds)`: the search's integer-division form of `step_at`.
    #[inline]
    pub fn step_after(&self, depart_offset_s: f64, seconds: f64) -> usize {
        let step = trunc_i64(floordiv(depart_offset_s + seconds, STEP_S));
        if step < 0 {
            return 0;
        }
        let last = self.n_steps as i64 - 1;
        (if step < last { step } else { last }) as usize
    }

    /// `depth_at(segment_id, step)`.
    #[inline]
    pub fn depth_at(&self, segment: u32, step: usize) -> f64 {
        match &self.depth[segment as usize] {
            Some(series) if !series.is_empty() => at(series, step),
            _ => 0.0,
        }
    }

    /// The `p_gt` series for a threshold, as `self.p_gt.get(float(threshold_cm))`.
    #[inline]
    pub fn p_for(&self, threshold_cm: f64) -> Option<&Series> {
        self.p_gt
            .iter()
            .find(|(t, _)| *t == threshold_cm)
            .map(|(_, s)| s)
    }

    /// `exceedance(segment_id, threshold_cm, step)`.
    pub fn exceedance(&self, segment: u32, threshold_cm: f64, step: usize) -> f64 {
        if let Some(by_segment) = self.p_for(threshold_cm) {
            if let Some(series) = &by_segment[segment as usize] {
                return at(series, step);
            }
            if self.depth[segment as usize].is_none() {
                return 0.0;
            }
        }
        if self.depth_at(segment, step) > threshold_cm {
            1.0
        } else {
            0.0
        }
    }

    /// `peak(segment_id)`.
    pub fn peak(&self, segment: u32) -> f64 {
        match &self.depth[segment as usize] {
            Some(series) if !series.is_empty() => {
                // Python's max(): the first of equal maxima, and NaN never wins a `>`.
                let mut best = series[0];
                for &v in &series[1..] {
                    if v > best {
                        best = v;
                    }
                }
                best
            }
            _ => 0.0,
        }
    }

    /// `time_of(step)`.
    pub fn time_of(&self, step: usize) -> DateTime {
        if step < self.times.len() {
            return self.times[step];
        }
        let last = *self.times.last().expect("a run has at least one step time");
        let minutes = STEP_MIN as i64 * (step as i64 - self.times.len() as i64 + 1);
        DateTime {
            utc_us: last.utc_us + minutes * 60 * crate::pytime::US_PER_S,
            offset_us: last.offset_us,
        }
    }
}

/// `series[step] if step < len(series) else series[-1]`.
///
/// An empty `p_gt` series makes Python raise `IndexError` (a 500); no run has one, and this
/// answers 0.0 rather than panicking a worker thread.
#[inline]
pub fn at(series: &[f64], step: usize) -> f64 {
    if step < series.len() {
        series[step]
    } else if let Some(&last) = series.last() {
        last
    } else {
        0.0
    }
}

/// Why a run could not be loaded, worded as the Python router words it.
#[derive(Debug)]
pub enum RunError {
    /// `FileNotFoundError`: the API answers 404 `no_run`.
    NotFound(String),
    /// A `run_id` that would escape `data/runs` (Python raises `ValueError`).
    BadRunId(String),
    /// A file that exists and cannot be read as a run.
    Invalid(String),
}

impl std::fmt::Display for RunError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RunError::NotFound(m) | RunError::BadRunId(m) | RunError::Invalid(m) => f.write_str(m),
        }
    }
}

fn load(path: &Path, graph: &RoadGraph) -> Result<SegmentDepths, RunError> {
    let name = path
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_default();
    let raw = std::fs::read(path.join("segments_wet.json"))
        .map_err(|e| RunError::Invalid(format!("{name}/segments_wet.json: {e}")))?;
    let wet: WetFile = serde_json::from_slice(&raw)
        .map_err(|e| RunError::Invalid(format!("{name}/segments_wet.json: {e}")))?;

    let depth_map = wet.depth_cm.unwrap_or_default();
    let n_steps = depth_map.values().map(Vec::len).max().unwrap_or(1);
    let mut times = Vec::new();
    for stamp in wet.valid_ts.unwrap_or_default() {
        match fromisoformat(&str_value(&stamp)) {
            Some(Parsed::Aware(dt)) => times.push(dt),
            _ => {
                return Err(RunError::Invalid(format!(
                "{name}/segments_wet.json has a step time that is not ISO 8601 with an offset: {}",
                str_value(&stamp)
            )))
            }
        }
    }
    if times.is_empty() {
        return Err(RunError::Invalid(format!(
            "{name}/segments_wet.json carries no step times; it cannot be routed against."
        )));
    }

    let n_seg = graph.segment_ids.len();
    let mut depth: Series = vec![None; n_seg];
    for (sid, series) in depth_map {
        if let Some(&j) = graph.segment_index.get(&sid) {
            depth[j as usize] = Some(series);
        }
    }

    let raw_p = wet.p_gt.unwrap_or_default();
    let has_exceedance = !raw_p.is_empty();
    let mut p_gt: Vec<(f64, Series)> = Vec::new();
    for (threshold, by_segment) in raw_p {
        let Some(t) = crate::pynum::parse_py_float(&threshold) else {
            return Err(RunError::Invalid(format!(
                "{name}/segments_wet.json has a p_gt threshold that is not a number: {threshold}"
            )));
        };
        let mut series: Series = vec![None; n_seg];
        for (sid, values) in by_segment {
            if let Some(&j) = graph.segment_index.get(&sid) {
                series[j as usize] = Some(values);
            }
        }
        p_gt.push((t, series));
    }

    let mut ensemble_n = 1;
    let mut rain = Vec::new();
    let meta_path = path.join("run.json");
    if meta_path.is_file() {
        let raw = std::fs::read(&meta_path)
            .map_err(|e| RunError::Invalid(format!("{name}/run.json: {e}")))?;
        let meta: RunFile = serde_json::from_slice(&raw)
            .map_err(|e| RunError::Invalid(format!("{name}/run.json: {e}")))?;
        // int(meta.get("ensemble_n", 1) or 1)
        ensemble_n = match meta.ensemble_n {
            Some(v) if crate::pynum::truthy(&v) => v
                .as_f64()
                .map(|f| f.trunc() as i64)
                .or_else(|| v.as_str().and_then(|s| s.trim().parse().ok()))
                .unwrap_or(1),
            _ => 1,
        };
        rain = meta.rain_aoi_mm_h.unwrap_or_default();
    }

    let run_id = match wet.run_id {
        Some(v) => str_value(&v),
        None => name,
    };
    Ok(SegmentDepths {
        run_id,
        times,
        depth,
        n_steps,
        ensemble_n,
        rain_aoi_mm_h: rain,
        p_gt,
        has_exceedance,
    })
}

/// Loads runs on demand and keeps the last few, invalidated by the file's mtime
/// (`forecast._load`'s `lru_cache(maxsize=4)` keyed on path and `st_mtime_ns`).
pub struct RunStore {
    pub data_dir: PathBuf,
    cache: Mutex<Vec<(PathBuf, SystemTime, Arc<SegmentDepths>)>>,
}

const CACHE_SIZE: usize = 8;

impl RunStore {
    pub fn new(data_dir: PathBuf) -> RunStore {
        RunStore {
            data_dir,
            cache: Mutex::new(Vec::new()),
        }
    }

    pub fn runs_dir(&self) -> PathBuf {
        self.data_dir.join("runs")
    }

    /// `latest_run_dir(city)`: the newest run for this city that carries `segments_wet.json`.
    pub fn latest_run_dir(&self, city_code: &str) -> Result<PathBuf, RunError> {
        let root = self.runs_dir();
        let prefix = if city_code.is_empty() {
            String::new()
        } else {
            format!("{city_code}-")
        };
        let mut names: Vec<(String, PathBuf)> = Vec::new();
        if let Ok(entries) = std::fs::read_dir(&root) {
            for entry in entries.flatten() {
                let p = entry.path();
                let name = entry.file_name().to_string_lossy().into_owned();
                if p.is_dir()
                    && (prefix.is_empty() || name.starts_with(&prefix))
                    && p.join("segments_wet.json").is_file()
                {
                    names.push((name, p));
                }
            }
        }
        // Python sorts Path objects, which on Windows compare case-insensitively.
        names.sort_by(|a, b| {
            let key = |s: &str| {
                if cfg!(windows) {
                    s.to_lowercase()
                } else {
                    s.to_string()
                }
            };
            key(&b.0).cmp(&key(&a.0))
        });
        names.into_iter().next().map(|(_, p)| p).ok_or_else(|| {
            RunError::NotFound(
                "No baked run to route against. Press Play on the replay, or run \
                 `make bake BUNDLE=MUM-2019-07-02`."
                    .into(),
            )
        })
    }

    /// `load_depths(run_id, city)`.
    pub fn load(
        &self,
        run_id: Option<&str>,
        city_code: &str,
        graph: &RoadGraph,
    ) -> Result<Arc<SegmentDepths>, RunError> {
        let path = match run_id {
            Some(id) => {
                if id.is_empty()
                    || id == "."
                    || id == ".."
                    || id.contains('/')
                    || id.contains('\\')
                    || id.contains('\0')
                {
                    return Err(RunError::BadRunId(format!(
                        "run_id must be a single path segment, got {}",
                        crate::pynum::repr_str(id)
                    )));
                }
                self.runs_dir().join(id)
            }
            None => self.latest_run_dir(city_code)?,
        };
        let wet = path.join("segments_wet.json");
        let meta = std::fs::metadata(&wet).ok().filter(|m| m.is_file());
        let Some(meta) = meta else {
            let name = path
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_default();
            return Err(RunError::NotFound(format!(
                "Run {name} has no segment forecast; it cannot be routed against."
            )));
        };
        let mtime = meta.modified().unwrap_or(SystemTime::UNIX_EPOCH);
        {
            let cache = self.cache.lock().expect("run cache lock");
            if let Some((_, _, hit)) = cache.iter().find(|(p, m, _)| *p == path && *m == mtime) {
                return Ok(Arc::clone(hit));
            }
        }
        let loaded = Arc::new(load(&path, graph)?);
        let mut cache = self.cache.lock().expect("run cache lock");
        cache.retain(|(p, _, _)| *p != path);
        cache.insert(0, (path, mtime, Arc::clone(&loaded)));
        cache.truncate(CACHE_SIZE);
        Ok(loaded)
    }
}
